Overview
-------------------------

Here we list links and references to the external software used in our analyses.

cutadapt
-------------------------

> "Cutadapt removes adapter sequences from high-throughput sequencing reads".
> Martin, EMBnet.journal, vol. 17, no. 1, p. 10, May 2011.
> https://github.com/marcelm/cutadapt

EPA-ng
-------------------------

> "EPA-ng: Massively Parallel Evolutionary Placement of Genetic Sequences".
> Barbera et al., bioRxiv, 2018.
> https://github.com/Pbdas/epa-ng

IQ-tree
-------------------------

> "IQ-TREE: A fast and effective stochastic algorithm for estimating maximum-likelihood phylogenies".
> Nguyen et al., Mol. Biol. Evol., vol. 32, no. 1, pp. 268–74, Jan. 2015.
> http://www.iqtree.org/

PaPaRa
-------------------------

> "PaPaRa 2.0: A Vectorized Algorithm for Probabilistic Phylogeny-Aware Alignment Extension".
> Berger et al., Heidelberg, 2012.
> https://sco.h-its.org/exelixis/web/software/papara/index.html

PEAR
-------------------------

> "PEAR: A fast and accurate Illumina Paired-End reAd mergeR".
> Zhang et al., Bioinformatics, vol. 30, no. 5, pp. 614–620, 2014.
> https://sco.h-its.org/exelixis/web/software/pear/

RAxML
-------------------------

> "RAxML version 8: a tool for phylogenetic analysis and post-analysis of large phylogenies".
> Stamatakis, Bioinformatics, vol. 30, no. 9, pp. 1312–1313, 2014.
> https://sco.h-its.org/exelixis/web/software/raxml/index.html

Sativa
-------------------------

> "Phylogeny-aware identification and correction of taxonomically mislabeled sequences".
> Kozlov et al., Nucleic Acids Res., vol. 44, no. 11, pp. 5022–5033, Jun. 2016.
> https://github.com/amkozlov/sativa

vsearch
-------------------------

> "VSEARCH: a versatile open source tool for metagenomics".
> Rognes et al., PeerJ, vol. 4, p. e2584, 2016.
> https://github.com/torognes/vsearch

